scripts/check_cmake_includes.py

```diff
@@ -0,0 +1,459 @@
+  defines: FrozenSet[str]
+
+
+class ConfigureFileParser:
+  """Parses a cmake `configure_file` file, enumerating its defines."""
+
+  def __init__(self, path: pathlib.Path) -> None:
+    """Initializes a new instance of this class.
+
+    Arguments:
+      path: The file to parse.
+    """
+    self.path = path
+
+  def parse(self) -> ConfigureFileParserParseResult:
+    """Parse the file.
+
+    Returns:
+      The data parsed from the file.
+    """
+    defines = set()
+
+    define_pattern = re.compile(r"#cmakedefine\s+(?P<name>[^\s]+)")
+
+    with self.path.open("rt", encoding="utf8") as f:
+      for line in f:
+        match = define_pattern.match(line)
+        if match:
+          defines.add(match.group("name"))
+
+    return ConfigureFileParserParseResult(defines=frozenset(defines))
+
+
+class CppFileParserResult(NamedTuple):
+  """Stores the result of parsing a C++ file."""
+
+  # The paths of the files included by #include.
+  includes: FrozenSet[str]
+
+  # The preprocessor variables used; the keys are the defines and the
+  # corresponding values are the line numbers.
+  defines_used: Mapping[str, int]
+
+
+class CppFileParser:
+  """Parser for C++ files to get the #include directives #define variables."""
+
+  def __init__(self, path: pathlib.Path) -> None:
+    """Initializes a new instance of this class.
+
+    Arguments:
+      path: The file to parse.
+    """
+    self.path = path
+
+  def parse(self) -> CppFileParserResult:
+    """Parse the file.
+
+    Returns:
+      The data parsed from the file.
+    """
+    includes = set()
+    defines_used = {}
+
+    include_pattern = re.compile(r"#(?i:include)\s+\"(?P<path>.+)\"$")
+    directive_pattern = re.compile(r"#\s*(?P<cmd>\w+)\s+(?P<var>\w+)")
+
+    with self.path.open("rt", encoding="utf8") as f:
+      line_number = 0
+      for line in f:
+        line_number += 1
+        include_match = include_pattern.match(line)
+        if include_match:
+          includes.add(include_match.group("path"))
+          continue
+
+        define_match = directive_pattern.match(line)
+        if define_match:
+          directive_name = define_match.group("cmd")
+          if directive_name.lower() == "define":
+            continue
+          define_value = define_match.group("var")
+          if define_value not in defines_used:
+            defines_used[define_value] = line_number
+
+    return CppFileParserResult(
+        includes=frozenset(includes),
+        defines_used=defines_used,
+    )
+
+
+class MissingInclude(NamedTuple):
+  """Stores information about a missing include."""
+
+  # The preprocessor variable that was used whose #include directive is missing.
+  define: str
+
+  # The string argument to the #include directive that was expected to be
+  # present for the define but was missing.
+  include: str
+
+  # The line number of the file at which the define was used.
+  line_number: int
+
+
+class RequiredIncludesChecker:
```

Comment:
If you kept a count of the files processed in here and added a summarize method that just wrote to stdout, you could eliminate the logger.